sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -871,28 +888,47 @@ pi_result piDevicesGet(pi_platform Platform, pi_device_type DeviceType,
     return PI_SUCCESS;
   }
 
-  // if devices are already captured in cache, return them from the cache.
+  // Return the devices from the cache.
+  uint32_t I = 0;
   for (const pi_device CachedDevice : Platform->PiDevicesCache) {
-    *Devices++ = CachedDevice;
+    if (I < NumEntries) {
+      *Devices++ = CachedDevice;
+      I++;
+    } else {
+      break;
+    }
   }
+
+  return PI_SUCCESS;
+}
+
+// Check the device cache and load it if necessary. The PiDevicesCacheMutex must
+// be locked before calling this function to prevent any synchronization issues.
+static pi_result populateDeviceCacheIfNeeded(pi_platform Platform) {
+
+  ze_driver_handle_t ZeDriver = Platform->ZeDriver;
+  uint32_t ZeDeviceCount = Platform->PiDevicesCache.size();
+  if (ZeDeviceCount == 0) {
+    ZE_CALL(zeDeviceGet(ZeDriver, &ZeDeviceCount, nullptr));
+  }
+
   if (!Platform->PiDevicesCache.empty()) {
     return PI_SUCCESS;
   }
 
   try {
     std::vector<ze_device_handle_t> ZeDevices(ZeDeviceCount);
     ZE_CALL(zeDeviceGet(ZeDriver, &ZeDeviceCount, ZeDevices.data()));
+    pi_device Device;
 
     for (uint32_t I = 0; I < ZeDeviceCount; ++I) {
-      if (I < NumEntries) {
-        Devices[I] = new _pi_device(ZeDevices[I], Platform);
-        pi_result Result = Devices[I]->initialize();
-        if (Result != PI_SUCCESS) {
-          return Result;
-        }
-        // save a copy in the cache for future uses.
-        Platform->PiDevicesCache.push_back(Devices[I]);
+      Device = new _pi_device(ZeDevices[I], Platform);
```

Comment:
There is no code to delete this pointer, but that bug existed even before your change.  One way to fix this is to use `unique_ptr`:

```
std::unique_ptr<_pi_device> Device(new _pi_device(ZeDevices[I], Platform);
[...]
Platform->PiDevicesCache.push_back(std::move(Device));
```

This assumes that `PiDevicesCache` is declared something like `std::vector<std::unique_ptr<pi_device>>`.

(This is all from memory, but I think this is mostly right. :-))

Reply:
Done.